airgun/session.py

```diff
@@ -261,7 +265,7 @@ def _prepare_browser(self, url):
             # Navigator
             self.navigator = Navigate(self.browser)
             self.navigator.dest_dict = navigator.dest_dict.copy()
-            if self._session_cookie is None:
+            if self._session_cookie is None and login_type is None:
                 self.login.login({
                     'username': self._user, 'password': self._password})
```

Comment:
The problem I have with that is that we pass `login_type` parameter with specific value, but we never care about this value. We over-expose information - we tell Airgun something, but we don't even expect it to act upon it.

I guess it really breaks down to how many tests with alternative login flow you plan, and what will be their goal.

Option 1
----------

If we want to tell Airgun about login method we will use, it naturally follows that we expect Airgun to **use** that login method.

On Airgun side, it would mean two changes:

1. Change Session `__init__` and `__call__` to support new parameter, say `login_type`
2. Let Airgun select login provider based on parameter presence/value:
```python
if self._session_cookie is not None:
    return  # session cookie logged us in already, we are done

login_provider = self.login
if self._login_type == "RHSSO":
    login_provider = self.rhsso_login

login_provider.login({'username': self._user, 'password': self._password})
```
(You could store username and password dict in variable instead, if that's more to your liking)

Then, in your test you would do:
```
with session(login_type="RHSSO"):
    session.user.search('')  # at this point, we are already logged in
```

The obvious drawback of that solution is that logging in is a little hidden from test perspective, while using alternative provider for logging in is really the core of what this test is about.

Option 2
---------------

If we don't want Airgun to perform logging in for us (because we want that explicitly in test), I suggest we don't over-complicate it by telling Airgun specific logging method. Just teach Airgun a way to communicate our real intent, which is "I will perform login myself, don't do that for me".

Then, new session parameter should be something like `log_in`, defaulting to `True`. When it is `False`, `self.login.login()` line is not executed.

------------

Again, the choice between these two breaks down to number of tests you plan to have and what will be their purpose. For small number of tests whose main idea is "this login method works", I would recommend option 2. If you already plan tests where alternative login method is just a step towards larger goals, I would recommend option 1.

Let me know if you want to discuss that outside of github.

Reply:
The login is an important part of authentication, but there will be tests which could have authentication and authorization. 
e.g. new rhsso user can log in? Is that user that right permissions not planned right now but in future definitely.
I like both the ideas, but I want to test should explicitly mention the login type/method. Because any new user would see and come to know what's happening instead of looking into airgun code. User should get at least come a level of understanding. Let's discuss in meeting I have some more doubts